lib/spirv/lib/src/instructions.dart

```diff
@@ -1,34 +1,306 @@
 part of spirv;
 
+class _BlockContext {
+  final StringBuffer out;
+  final int indent;
+
+  /// The most local merge block id, or zero.
+  final int merge;
+
+  /// The most local continue block id, or zero.
+  final int continueBlock;
+
+  /// The most local loop-construct header block id.
+  final int loopHeader;
+
+  /// The most local loop-construct merge block id.
+  /// This is different from [merge] when the context is inside an if-statement
+  /// inside of a for-loop, for example.
+  final int loopMerge;
+
+  _BlockContext({
+    required this.out,
+    required this.indent,
+    this.merge = 0,
+    this.continueBlock = 0,
+    this.loopHeader = 0,
+    this.loopMerge = 0,
+  });
+
+  /// Return a new [_BlockContext] that is a copy of the current [_BlockContext]
+  /// with an increased indent and any parameters specified here overwritten.
+  _BlockContext child({
+    int? merge,
+    int? continueBlock,
+    int? loopHeader,
+    int? loopMerge,
+  }) => _BlockContext(
+    out: out,
+    indent: indent+1,
+    merge: merge ?? this.merge,
+    continueBlock: continueBlock ?? this.continueBlock,
+    loopHeader: loopHeader ?? this.loopHeader,
+    loopMerge: loopMerge ?? this.loopMerge,
+  );
+
+  void writeIndent() {
+    for (int i = 0; i < indent; i++) {
+      out.write('  ');
+    }
+  }
+}
+
 class _Block {
   List<_Instruction> instructions = <_Instruction>[];
 
   void add(_Instruction i) {
     instructions.add(i);
   }
 
-  void writeIndent(StringBuffer out, int indent) {
-    for (int i = 0; i < indent; i++) {
-      out.write('  ');
+  final int id;
+  final _Function function;
+  final _Transpiler transpiler;
+
+  _Block(this.id, this.function, this.transpiler);
+
+  // control flow
+  int branch = 0;
+  int mergeBlock = 0;
+  int condition = 0;
+  int truthyBlock = 0;
+  int falseyBlock = 0;
+
+  // structured loop
+  _Store? loopInitializer;
+  int continueBlock = 0;
+
+  /// true if this block has been processed by [liftLoopVariables].
+  bool scanned = false;
+
+  void writeContinue(_BlockContext ctx) {
+    bool written = false;
+    for (final _Instruction inst in instructions) {
```

Comment:
Can you use `instructions.whereType<_Store>()`?  https://api.dart.dev/dev/2.17.0-108.0.dev/dart-core/Iterable/whereType.html